### PR TITLE
CNF-14531: Tests for loss of GNSS signal and holdover

### PR DESCRIPTION
**Pull request by nishant-parekh**

### CODE REVIEW

File: test/conformance/serial/ptp.go [hunk 1]
```diff
@@ -1229,49 +1088,163 @@ var _ = Describe("["+strings.ToLower(DesiredMode.String())+"-serial]", Serial, f
...
+					// Disconnect GNSS signal using
+					//```ubxtool -P 29.20 -p CFG-MSG,0xf0,4,0```
+					pods.ExecCommand(client.Client, fullConfig.DiscoveredClockUnderTestPod,
+						pkg.PtpContainerName, []string{"ubxtool", "-P", "29.20", "-p", "CFG-MSG,0xf0,4,0"})
+
+					// Verify state using metrics
+					// openshift_ptp_clock_class (7 in spec/ 140 out of spec)
+					checkClockClassState(fullConfig, "7")
+					// openshift_ptp_clock_state (GM,DPLL,ts2phc => 2 holdover)
+					checkClockState(fullConfig, "2")
```
Comment (aneeshkp) [inline test/conformance/serial/ptp.go:1125]: check only T-GM  State

File: test/conformance/serial/ptp.go [hunk 2]
```diff
@@ -1229,49 +1088,163 @@ var _ = Describe("["+strings.ToLower(DesiredMode.String())+"-serial]", Serial, f
...
+
+					// Recover GNSS Connection signal using
+					//```ubxtool -P 29.20 -p CFG-MSG,0xf0,4,1```
+					pods.ExecCommand(client.Client, fullConfig.DiscoveredClockUnderTestPod,
+						pkg.PtpContainerName, []string{"ubxtool", "-P", "29.20", "-p", "CFG-MSG,0xf0,4,1"})
+
+					// Verify state using metrics
+					//openshift_ptp_clock_class (6)
+					checkClockClassState(fullConfig, "6")
```
Comment (aneeshkp) [inline test/conformance/serial/ptp.go:1134]: use enum for clock class

File: test/conformance/serial/ptp.go [hunk 3]
```diff
@@ -1229,49 +1088,163 @@ var _ = Describe("["+strings.ToLower(DesiredMode.String())+"-serial]", Serial, f
...
+
+				})
+
+				By("disconnecting the GNSS Signal to test Holdover capability WPC GM", func() {
+					// Check Stability of system using metrics
+					checkStabilityOfWPCGMUsingMetrics(fullConfig)
+
+					// Disconnect GNSS signal using
+					//```ubxtool -P 29.20 -p CFG-MSG,0xf0,4,0```
+					pods.ExecCommand(client.Client, fullConfig.DiscoveredClockUnderTestPod,
```
Comment (aneeshkp) [inline test/conformance/serial/ptp.go:1146]: This won't work , use cold boot

File: test/conformance/serial/ptp.go [hunk 4]
```diff
@@ -1229,49 +1088,163 @@ var _ = Describe("["+strings.ToLower(DesiredMode.String())+"-serial]", Serial, f
...
+
+					// Verify state using metrics
+					// openshift_ptp_clock_class (7 in spec/ 140 out of spec)
+					checkClockClassState(fullConfig, "7")
+					// openshift_ptp_clock_state (GM,DPLL,ts2phc => 2 holdover)
+					checkClockState(fullConfig, "2")
+
+					// Recover GNSS Connection signal using
+					//```ubxtool -P 29.20 -p CFG-MSG,0xf0,4,1```
+					pods.ExecCommand(client.Client, fullConfig.DiscoveredClockUnderTestPod,
```
Comment (aneeshkp) [inline test/conformance/serial/ptp.go:1129]: Same here

File: test/conformance/serial/ptp.go [hunk 5]
```diff
@@ -1229,49 +1088,163 @@ var _ = Describe("["+strings.ToLower(DesiredMode.String())+"-serial]", Serial, f
...
+					//```ubxtool -P 29.20 -p CFG-MSG,0xf0,4,0```
+					pods.ExecCommand(client.Client, fullConfig.DiscoveredClockUnderTestPod,
+						pkg.PtpContainerName, []string{"ubxtool", "-P", "29.20", "-p", "CFG-MSG,0xf0,4,0"})
+
+					// Verify state using metrics
+					// Wait until state updates from
+					// locked => holdover inspec => holdover out of spec => freerun
+
+					// openshift_ptp_clock_class (7 in spec/ 140 out of spec)
+					checkClockClassState(fullConfig, "7")
```
Comment (aneeshkp) [inline test/conformance/serial/ptp.go:1154]: use enum

File: test/conformance/serial/ptp.go [hunk 6]
```diff
@@ -1229,49 +1088,163 @@ var _ = Describe("["+strings.ToLower(DesiredMode.String())+"-serial]", Serial, f
...
+					// openshift_ptp_clock_state (GM,DPLL,ts2phc => 2 holdover)
+					checkClockState(fullConfig, "2")
+
+					time.Sleep(2 * time.Minute)
+
+					// openshift_ptp_clock_class (248 freerun after 2 mins)
+					checkClockClassState(fullConfig, "248")
+					// openshift_ptp_clock_state (GM,DPLL,ts2phc 0 freerun)
+					checkClockState(fullConfig, "0")
```
Comment (aneeshkp) [inline test/conformance/serial/ptp.go:1163]: enum

File: test/conformance/serial/ptp.go [hunk 7]
```diff
@@ -1229,49 +1088,163 @@ var _ = Describe("["+strings.ToLower(DesiredMode.String())+"-serial]", Serial, f
...
+
+					//Recover GNSS Connection signal using
+					//```ubxtool -P 29.20 -p CFG-MSG,0xf0,4,1```
+					pods.ExecCommand(client.Client, fullConfig.DiscoveredClockUnderTestPod,
+						pkg.PtpContainerName, []string{"ubxtool", "-P", "29.20", "-p", "CFG-MSG,0xf0,4,1"})
```
Comment (aneeshkp) [inline test/conformance/serial/ptp.go:1168]: Doesn't work

File: test/conformance/serial/ptp.go [hunk 8]
```diff
@@ -1229,49 +1088,163 @@ var _ = Describe("["+strings.ToLower(DesiredMode.String())+"-serial]", Serial, f
...
+
+					//Verify state using metrics
+					//openshift_ptp_clock_class (6)
+					checkClockClassState(fullConfig, "6")
```
Comment (aneeshkp) [inline test/conformance/serial/ptp.go:1172]: enum

File: test/conformance/serial/ptp.go [hunk 9]
```diff
@@ -1229,49 +1088,163 @@ var _ = Describe("["+strings.ToLower(DesiredMode.String())+"-serial]", Serial, f
...
+					Skip("To be implemented")
+				})
 
+			})
 		})
 	})
 })
 
+func checkStabilityOfWPCGMUsingMetrics(fullConfig testconfig.TestConfig) {
+	checkProcessStatus(fullConfig, "1")
```
Comment (aneeshkp) [inline test/conformance/serial/ptp.go:1240]: use all existing enum here

File: test/conformance/serial/ptp.go [hunk 10]
```diff
@@ -1385,3 +1358,238 @@ func clockStateByProcesses(input string) (map[string]bool, error) {
...
+		"phc2ys process status not detected")
+
+	time.Sleep(10 * time.Second)
+	buf, _, _ := pods.ExecCommand(client.Client, fullConfig.DiscoveredClockUnderTestPod, pkg.PtpContainerName, []string{"curl", pkg.MetricsEndPoint})
+	ret, err := processRunning(buf.String(), state)
+	Expect(err).To(BeNil())
+	Expect(ret["phc2sys"]).To(BeTrue(), fmt.Sprintf("Expected phc2sys to be  %s for GM", state))
+	Expect(ret["ptp4l"]).To(BeTrue(), fmt.Sprintf("Expected ptp4l to be  %s for GM", state))
+	Expect(ret["ts2phc"]).To(BeTrue(), fmt.Sprintf("Expected ts2phc to be  %s for GM", state))
+	Expect(ret["gpspipe"]).To(BeTrue(), fmt.Sprintf("Expected gpspipe to be %s for GM", state))
```
Comment (aneeshkp) [inline test/conformance/serial/ptp.go:1391]: this won;t work yet

**Comment by nishant-parekh** [comment]
@aneeshkp I have fixed all the issues you listed here in this PR:
https://github.com/openshift/ptp-operator/pull/548
This covers all the additional tests in the as per our test plan.

**Pull request merged**